### PR TITLE
add configuration for normalize translation chain

### PR DESCRIPTION
**Pull request by flug**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -4,7 +4,8 @@
     "keywords": ["enum", "enumeration"],
     "require": {
         "greg0ire/enum": "~1.0",
-        "symfony/form": "~2.3"
+        "symfony/form": "~2.3", 
+        "symfony/dependency-injection": "^2.7"
```
Comment (greg0ire) [inline composer.json]: Can you add this dependency _before_ the symfony/form dep ? Otherwise the diff is strange.

File: README.md [hunk 2]
```diff
@@ -6,7 +6,7 @@ yet but :
 - it has a dependency on symfony/form
 - it could become a real bundle someday, if something needs to be configured
 
-[![Build Status][2]](https://travis-ci.org/greg0ire/enum-bundle)
+[![Build Status][3]](https://travis-ci.org/greg0ire/enum-bundle)
```
Comment (greg0ire) [inline README.md]: WTF ?

File: src/Bundle/BootEnum.php [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?php
+namespace Greg0ire\EnumBundle\Bundle;
+
+class BootEnum extends \Greg0ire\EnumBundle\BaseEnum
+{
+    public static function inject($parameters)
+    {
+        parent::inject($parameters);
+    }
+}
```
Comment (roukmoute) [inline src/Bundle/BootEnum.php]: missing carriage return

File: src/DependencyInjection/Greg0ireEnumExtension.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+     */
+    public function load(array $configs, ContainerBuilder $container)
+    {
+        $configuration = new Configuration();
+        $config = $this->processConfiguration($configuration, $configs);
+
+        $container->setParameter('greg0ire_enum.chain', $config['chain']);
+
+    }
+}
```
Comment (roukmoute) [inline src/DependencyInjection/Greg0ireEnumExtension.php]: Missing carriage return

File: README.md [hunk 5]
```diff
@@ -15,8 +15,8 @@ yet but :
 ## Usage
 
 The bundle provides its own `BaseEnum` class. It inherits from `greg0ire/enum`'s
-`BaseEnum` class and provides an additional method, `getChoices()`, which
-is meant to be used as value for the `choices` option of a choice widget.
+`BaseEnum` class and provides an additional method, `getChoiceList()`, which
+is meant to be used as value for the `choice_list` option of a choice widget.
```
Comment (greg0ire) [inline README.md]: WTF ?

File: README.md [hunk 6]
```diff
@@ -45,21 +45,67 @@ class MyType
...
+```php
+<?php
+
+class MyType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder->add(
+            'aspect_ratio',
+            'choice',
```
Comment (greg0ire) [inline README.md]: This should not be in your P.R.

File: README.md [hunk 7]
```diff
@@ -45,21 +45,67 @@ class MyType
...
             array('choices' => ColorType::getChoices('color_type_%s'))
         );
     }
 }
 ```
 
+Possibility of configuration of the translation chain
```
Comment (greg0ire) [inline README.md]: Please add a "Configuration" section, like this : 

```
## Configuration
```

File: README.md [hunk 8]
```diff
@@ -45,21 +45,67 @@ class MyType
...
 - `color_type_color`
 - `color_type_colorized`
 
-The argument to `getChoices()` is optional, and the value will be used directly
-as a label should you choose not to specify it.  This makes sense if you decide
-to have a translation catalogue just for your enumeration.
+The argument to both `getChoiceList()` and `getChoices()` is optional, and the
+value will be used directly as a label should you choose not to specify it.
+This makes sense if you decide to have a translation catalogue just for your
+enumeration.
```
Comment (greg0ire) [inline README.md]: This should not be in your P.R.

File: README.md [hunk 9]
```diff
@@ -45,21 +45,67 @@ class MyType
...
+    /** other bundles */
+    new Greg0ire\EnumBundle\Greg0ireEnumBundle(),
+
+);
+
+```
+
+In ```config.yml``` add :
+
+you can choose between **underscore** or **camelize**
```
Comment (greg0ire) [inline README.md]: Please remove this and add it as a comment

File: README.md [hunk 10]
```diff
@@ -45,21 +45,67 @@ class MyType
...
+
+```yml
+
+greg0ire_enum:
+    chain: underscore
```
Comment (greg0ire) [inline README.md]:
```
    chain: underscore # underscore | camelize
```

File: composer.json [hunk 11]
```diff
@@ -4,7 +4,8 @@
     "keywords": ["enum", "enumeration"],
     "require": {
         "greg0ire/enum": "~1.0",
-        "symfony/form": "~2.3"
+        "symfony/form": "~2.3,<2.7",
+        "symfony/dependency-injection": "^2.7"
```
Comment (greg0ire) [inline composer.json]: Why `2.7` ?

File: src/BaseEnum.php [hunk 12]
```diff
@@ -3,9 +3,13 @@
 namespace Greg0ire\EnumBundle;
 
 use Greg0ire\Enum\BaseEnum as LibraryEnum;
+use Symfony\Component\DependencyInjection\Container;
 
 abstract class BaseEnum extends LibraryEnum
 {
+
+    private static $parameters = null;
```
Comment (greg0ire) [inline src/BaseEnum.php]: There should be a better name than `parameters`, don't you think ? `formatter`, maybe ?

File: src/BaseEnum.php [hunk 13]
```diff
@@ -28,12 +32,25 @@ public static function getLabelArray($values, $pattern)
...
+
+    public static function prepareChain($pattern, $element)
+    {
+        $formatChain = BaseEnum::$parameters;
+
+        if ($formatChain == null) {
+            return sprintf($pattern, $element);
+        }
+
+        return Container::$formatChain(sprintf($pattern, $element));
```
Comment (greg0ire) [inline src/BaseEnum.php]: `Container` ?!?

File: src/Bundle/BootEnum.php [hunk 14]
```diff
@@ -0,0 +1,10 @@
+<?php
+namespace Greg0ire\EnumBundle\Bundle;
+
+class BootEnum extends \Greg0ire\EnumBundle\BaseEnum
+{
+    public static function inject($parameters)
+    {
+        parent::inject($parameters);
```
Comment (greg0ire) [inline src/Bundle/BootEnum.php]: so, this does nothing more than its parent ?

File: src/DependencyInjection/Configuration.php [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+     * {@inheritdoc}
+     */
+    public function getConfigTreeBuilder()
+    {
+        $treeBuilder = new TreeBuilder();
+        $rootNode = $treeBuilder->root('greg0ire_enum');
+
+        $rootNode->children()
+            ->enumNode('chain')
+                ->values(['underscore', 'camelize'])
```
Comment (greg0ire) [inline src/DependencyInjection/Configuration.php]: please add a call to `info()` to document this parameter

File: src/DependencyInjection/Configuration.php [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function getConfigTreeBuilder()
+    {
+        $treeBuilder = new TreeBuilder();
+        $rootNode = $treeBuilder->root('greg0ire_enum');
+
+        $rootNode->children()
+            ->enumNode('chain')
```
Comment (greg0ire) [inline src/DependencyInjection/Configuration.php]: This is not a chain, this is a format. A translation key format. Please name this parameter accordingly.

File: src/DependencyInjection/Configuration.php [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+namespace Greg0ire\EnumBundle\DependencyInjection;
+
+use Symfony\Component\Config\Definition\Builder\TreeBuilder;
+use Symfony\Component\Config\Definition\ConfigurationInterface;
+
+/**
+ * This is the class that validates and merges configuration from your app/config files
+ *
+ * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html#cookbook-bundles-extension-config-class}
+ */
```
Comment (greg0ire) [inline src/DependencyInjection/Configuration.php]: Please remove this comment.

File: src/DependencyInjection/Configuration.php [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+        $rootNode = $treeBuilder->root('greg0ire_enum');
+
+        $rootNode->children()
+            ->enumNode('chain')
+                ->values(['underscore', 'camelize'])
+                ->defaultNull()
+            ->end();
+        // Here you should define the parameters that are allowed to
+        // configure your bundle. See the documentation linked above for
+        // more information on that topic.
```
Comment (greg0ire) [inline src/DependencyInjection/Configuration.php]: Please remove this comment.

File: src/DependencyInjection/Greg0ireEnumExtension.php [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function load(array $configs, ContainerBuilder $container)
+    {
+        $configuration = new Configuration();
+        $config = $this->processConfiguration($configuration, $configs);
+
+        $container->setParameter('greg0ire_enum.chain', $config['chain']);
```
Comment (greg0ire) [inline src/DependencyInjection/Greg0ireEnumExtension.php]: Why not inject the callable directly in the `BaseEnum` class ? This class is not built with the DIC, so there is no point in creating a container parameter, is there ?

File: src/Bundle/BootEnum.php (hunk 3 above)
Comment (greg0ire) [inline src/Bundle/BootEnum.php]: Good catch @roukmoute ! @flug, please configure your editor to add a "newline" character add the end of each file. vim does that by default, phpstorm can be configured to do so.

File: src/DependencyInjection/Greg0ireEnumExtension.php [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+use Symfony\Component\Config\FileLocator;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Loader;
+use Symfony\Component\HttpKernel\DependencyInjection\Extension;
+
+/**
+ * This is the class that loads and manages your bundle configuration
+ *
+ * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html}
+ */
```
Comment (greg0ire) [inline src/DependencyInjection/Greg0ireEnumExtension.php]: Please remove this comment.

File: src/Greg0ireEnumBundle.php [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+
+use Greg0ire\EnumBundle\Bundle\BootEnum;
+use Symfony\Component\HttpKernel\Bundle\Bundle;
+
+class Greg0ireEnumBundle extends Bundle
+{
+    public function boot()
+    {
+        if ($this->container->hasParameter('greg0ire_enum.chain')) {
+            BootEnum::inject($this->container->getParameter('greg0ire_enum.chain'));
```
Comment (greg0ire) [inline src/Greg0ireEnumBundle.php]: Why not use directly `BaseEnum::inject` ?

File: src/BaseEnum.php [hunk 22]
```diff
@@ -28,12 +32,25 @@ public static function getLabelArray($values, $pattern)
...
+    }
+
+    public static function inject($parameters)
```
Comment (greg0ire) [inline src/BaseEnum.php]: This should be called `setParameters()`

**Comment by greg0ire** [comment]
This P.R. is a good idea, but I think it would be even better if people could use **any service they want** to do the formatting. You should create an interface, and say that people should implement it if they want to use it as a formatter.

**Pull request closed without merging**